saml service provider resource request validation

flyinghermit opened the pull request:
This PR addresses two issues related to the saml idp service provider resource access request:
1. Updates `[RequestValidator].roleAllowsResource` method to  use `CheckAccessToSAMLIdP` instead of `checkAccess` for SAML resource. `CheckAccessToSAMLIdP` itself wraps `checkAccess` method but does additional validation such as idp denied at role or cluster level. 
2. Updates `[RequestValidator].validate` to precheck user access to the requested SAML resource using combination of (requested roles + standing roles). SAML resource has different rbac behavior between role version `v8` and `v7-`. If a user has both version of the role, has been configured to request access with `v7` role but `v8` role does not permit access, user access request passes validation but access will be denied at the time of login. This new precheck will provide early feedback for the user. 

Both of these issue would otherwise give a false impression as user may successfully create and assume access request to sam resource, but access may be denied at the time of login. The main motivation for the change is to have saml app implicitly added when identity center account is requested for access https://github.com/gravitational/teleport/pull/61758

changelog: Fixed an issue with the `saml_idp_service_provider` resource access request validator which was missing full resource access validation similar to what is checked during time of login.

flyinghermit committed:
updated access request validator to check user access to saml resource using existing+requested roleset

flyinghermit committed:
Merge branch 'master' into sshah/saml-app-request-validator

flyinghermit committed:
fix test: update expected roles TestPruneMappedSearchAs/foreign_resource

flyinghermit committed:
update commentry and move checker below requested role checker

flyinghermit commented on lib/services/access_request.go:
```diff
@@ -2403,7 +2412,15 @@ func (m *RequestValidator) pruneResourceRequestRoles(
-				resourcesStr, roles, kubeResourceMatcher.Unmatched())
+				resourcesStr, requestedRoles, kubeResourceMatcher.Unmatched())
+		}
+		if resource.GetKind() == types.KindSAMLIdPServiceProvider {
```
Arguably abusing `pruneResourceRequestRoles` method to do many things but this is the place where roles and resource are listed and looped through and seemed like best place to slip this block.

flyinghermit committed:
use slices.Concat over append

flyinghermit committed:
prune legacy role for saml resource

flyinghermit committed:
Merge branch 'master' into sshah/saml-app-request-validator

flyinghermit commented:
Friendly ping for review please

flyinghermit committed:
Merge branch 'master' into sshah/saml-app-request-validator

nklaassen commented on lib/services/access_request.go:
```diff
@@ -2474,7 +2530,22 @@ func (m *RequestValidator) roleAllowsResource(
+			// legacy roles grant implicit access to saml resource
+			// and can be removed from the reqeust.
+			return false, nil
+		}
```
I'm trying to understand how this works, is it true that users with NEVER want to request a v7 role to access a SAMLIdPServiceProvider? I guess the behaviour is that if the user has a single v8+ role then a v7 role will never grant access, and if the user only has v7 roles then they already should have access (unless explicitly denied).

nklaassen commented on lib/services/access_request.go:
```diff
@@ -2324,15 +2330,15 @@ func (m *RequestValidator) pruneResourceRequestRoles(
 	resourceIDs []types.ResourceID,
 	loginHint string,
-	roles []string,
+	requestedRoles []string,
```
I sort of disagree with renaming this to `requestedRoles` since nobody has requested them at this point, this is just the full list of candidate roles that it's possible to request. You could name it `candidateRoles` if you prefer

nklaassen commented on lib/services/access_request.go:
```diff
@@ -2408,7 +2423,15 @@ func (m *RequestValidator) pruneResourceRequestRoles(
+			// resource even after assuming roles granted by this access request.
+			if err := canAccessSaml(localClusterName, roleSet, resource, m.userState.GetTraits(), matchers...); err != nil {
+				return nil, trace.Wrap(err)
+			}
```
This seems like it would return an error if it is merely possible to request a role that would not allow access to the requested SAMLIdPServiceProvider, where the desired behaviour would be for that role to just be filtered out and not requested

nklaassen commented on lib/services/access_request_test.go:
```diff
@@ -1956,6 +2012,128 @@ func TestPruneMappedRoles(t *testing.T) {
+			},
+		},
+		{
+			desc: "request fails if v8 role exists and does not grant access",
```
can we have a testcase where it's possible to request 2 v8 roles, one which grants access to the app and one which does not, and make sure it requests the correct role? I don't think the code you have right now will work for this case, but I think it should

smallinsky commented on lib/services/access_request.go:
```diff
@@ -1670,11 +1670,17 @@ func (m *RequestValidator) getRequestableRoles(ctx context.Context, identity tls
+		case types.KindSAMLIdPServiceProvider:
+			if err = accessChecker.CheckAccessToSAMLIdP(resource, nil /*AuthPreference*/, AccessState{MFAVerified: true}); err != nil {
+				filteredResources = append(filteredResources, resource)
+			}
```
At first glance this looks like sec bug where the rescues is added when resource is added when the err != nil -> Access Denied 

> // Filter out resources the user requested but doesn't have access to.

flyinghermit replied:
This was addressed in https://github.com/gravitational/teleport/pull/61204/commits/91179055625561d87c09fe35614aaa95c04bc9b1

flyinghermit committed:
refactor:
- revert pruneResourceRequestRoles changes
- move saml requested+standing role validator to validate method

flyinghermit committed:
test:
- check expectRole matches
- add test case to demonstrate saml app requested with two v8 roles passes with valid role

flyinghermit committed:
Merge branch 'master' into sshah/saml-app-request-validator

flyinghermit commented on lib/services/access_request_test.go:
```diff
@@ -1956,6 +2012,128 @@ func TestPruneMappedRoles(t *testing.T) {
+		requestResourceIDs []types.ResourceID
+		errorAssertion     require.ErrorAssertionFunc
+		overrideRoles      []string
+		expectRoles        []string
```
Thanks for that, updated test to check expectRoles https://github.com/gravitational/teleport/pull/61204/commits/794d39573a414ca1dc4be7c9ef8ba0b78b50b998

flyinghermit committed:
fix: requestable role filtered on inaccessible resource for CalculateAccessCapabilities

smallinsky commented on lib/services/access_request.go:
```diff
@@ -2582,6 +2610,68 @@ func getKubeResourcesFromResourceIDs(resourceIDs []types.ResourceID, clusterName
+		); err != nil {
+			if trace.IsAccessDenied(err) {
+				return trace.BadParameter(
+					`no roles configured in the "search_as_roles" for this user allow `+
```
How a user can issue a JIT Access Request with some requestedResourceIDs  set and then get the errro that he don't have the `search_as_roles` ? 


If user don't have any role `search_as_roles` my thinking is that we should not be able to issue any JIT access request because  it don't have permission to list any resources.
